File: LUA/encryptor.lua
local openssl = require("openssl")
local lfs = require("lfs")

local function generateKey()
    return openssl.random(32)
end

local function padData(data)
    local blockSize = 16
    local paddingLength = blockSize - (#data % blockSize)
    return data .. string.rep(string.char(paddingLength), paddingLength)
end

local function encryptFile(filePath, key)
    local iv = openssl.random(16)
    local cipher = openssl.cipher.get('aes-256-cbc')
    
    local file = io.open(filePath, "rb")
    local data = file:read("*all")
    file:close()
    
    local paddedData = padData(data)
    local encryptedData = cipher:encrypt(paddedData, key, iv)
    
    file = io.open(filePath, "wb")
    file:write(iv .. encryptedData)
    file:close()
end

local function processFolder(folderPath, key)
    for file in lfs.dir(folderPath) do
        if file ~= "." and file ~= ".." then
            local filePath = folderPath .. '/' .. file
            local attr = lfs.attributes(filePath)
            if attr.mode == "file" then
                encryptFile(filePath, key)
            end
        end
    end
end

local function main()
    local folderPath = "E:/Test/Folder"
    local parentPath = folderPath:match("(.+)/")
    local keyFilePath = parentPath .. "/key.txt"
    
    local key = generateKey()
    local keyFile = io.open(keyFilePath, "wb")
    keyFile:write(key)
    keyFile:close()
    
    processFolder(folderPath, key)
end

main()

File: Ruby/encryptor.rb
require 'openssl'
require 'securerandom'
require 'fileutils'

def generate_key
  SecureRandom.random_bytes(32)
end

def pad_data(data)
  block_size = 16
  padding_length = block_size - (data.bytesize % block_size)
  data + padding_length.chr * padding_length
end

def encrypt_file(file_path, key)
  iv = SecureRandom.random_bytes(16)
  cipher = OpenSSL::Cipher::AES256.new(:CBC)
  cipher.encrypt
  cipher.key = key
  cipher.iv = iv

  data = File.read(file_path, mode: 'rb')
  padded_data = pad_data(data)
  encrypted_data = cipher.update(padded_data) + cipher.final

  File.write(file_path, iv + encrypted_data, mode: 'wb')
end

def process_folder(folder_path, key)
  Dir.glob("#{folder_path}/*").each do |file_path|
    if File.file?(file_path)
      encrypt_file(file_path, key)
    end
  end
end

def main
  folder_path = "E:/Test/Folder"
  parent_path = File.dirname(folder_path)
  key_file_path = File.join(parent_path, "key.txt")

  key = generate_key
  File.write(key_file_path, key, mode: 'wb')

  process_folder(folder_path, key)
end

main
